snr.py

```diff
@@ -0,0 +1,287 @@
+        Total photons per pixel due to the background/sky with units of
+        electrons/pixel.
+        Default is 0 * astropy.units.electron / astropy.units.pixel.
+    darkcurrent : `~astropy.units.Quantity`, optional
+        Total electrons per pixel due to the dark current with units
+        of electrons/pixel.
+        Default is 0 * astropy.units.electron / astropy.units.pixel.
+    readnoise : `~astropy.units.Quantity`, optional
+        Electrons per pixel from the read noise with units of electrons/pixel.
+        Default is 0 * astropy.units.electron / astropy.units.pixel.
+    gain : `~astropy.units.Quantity`, optional
+        Gain of the CCD with units of electrons/ADU. Default is
+        1 * astropy.units.electron / astropy.units.adu such that the
+        contribution to the error due to the gain is assumed to be small.
+    ad_err : `~astropy.units.Quantity`, optional
+        An estimate of the 1 sigma error within the A/D converter with units of
+        adu/pixel. Default is set to
+        0.289 * astropy.units.adu / astropy.units.pixel
+        (Merline & Howell, 1995).
+
+    Returns
+    -------
+    sn : `~astropy.units.Quantity`
+        The signal to noise ratio of the given observation in untis of
+        sqrt(electrons).
+    """
+    readnoise = get_shotnoise(readnoise)
+    gain_err = get_shotnoise(gain * ad_err)
+
+    pixel_terms = npix * (1 + npix / n_background)
+    detector_noise = (background + darkcurrent +
+                      readnoise ** 2 + gain_err ** 2)
+
+    sn = counts / np.sqrt(counts + pixel_terms * detector_noise)
+
+    return sn
+
+
+def get_shotnoise(detector_property):
+    """
+    Returns the shot noise (i.e. non-Poissonion noise) in the correct
+    units.
+    """
+    return detector_property.value * np.sqrt(1 * u.electron / u.pixel)
+
+
+def t_with_small_errs(t, background_rate, darkcurrent_rate, gain_err,
+                      readnoise, countrate, npix, n_background):
+    """
+    Returns the full expression for the exposure time including the
+    contribution to the noise from the background and the gain.
+    """
+    if type(t) != 'astropy.units.quantity.Quantity':
+        t = t * u.s
+
+    detector_noise = (background_rate * t + darkcurrent_rate * t +
+                      gain_err ** 2 + readnoise ** 2)
+    radicand = countrate * t + (npix * (1 + npix/n_background) *
+                                detector_noise)
+
+    return countrate * t / np.sqrt(radicand)
+
+
+@u.quantity_input(snr=np.sqrt(1 * u.electron),
+                  countrate=u.electron / u.s,
+                  npix=u.pixel,
+                  n_background=u.pixel,
+                  background_rate=u.electron / u.pixel / u.s,
+                  darkcurrent_rate=u.electron / u.pixel / u.s,
+                  readnoise=u.electron / u.pixel,
+                  gain=u.electron / u.adu,
+                  ad_err=u.adu / u.pixel)
+def exposure_time_from_snr(snr, countrate,
```

Comment:
I think the name we were converging towards was `exptime_from_snr_howell2000`, or similar.